src/dbnode/x/encoding/shallow_cloner.go

```diff
@@ -0,0 +1,161 @@
+	}
+
+	iterCopy := s.pools.SeriesIterator().Get()
+	iterCopy.Reset(enc.SeriesIteratorOptions{
+		ID:             s.pools.ID().Clone(iter.ID()),
+		Tags:           iter.Tags().Duplicate(),
+		StartInclusive: iter.Start(),
+		EndExclusive:   iter.End(),
+		Replicas:       replicaCopies,
+	})
+
+	return iterCopy, nil
+}
+
+const (
+	_initReadersSize = 5
+)
+
+func (s shallowCloner) cloneMultiReaderIterator(
+	iter enc.MultiReaderIterator,
+) (enc.MultiReaderIterator, error) {
+	// TODO: pool these slice allocations
+	blockReaderCopies := make([][]xio.BlockReader, 0, _initReadersSize)
+
+	// nb: the implementation below requires iteration of the blocksIters
+	// underlying the MultiReaderIterator. While we make sure to reset the
+	// state of the iterator back to what it was originally, one important
+	// caveat to callout: this iteration is not-threadsafe. i.e. if copies
+	// are going to be made by different go-routines, the synchronization
+	// has to be done at a level above this.
+	readers := iter.Readers()
+	initIdx := readers.Index()
+
+	// nb: we start by assuming next=true for the first pass through the readersIterator
+	// as the multiReaderiterator already calls Next() on the readersIterator when
+	// it acquires the iter.
+	for next := true; next; next = readers.Next() {
+		currLen, currStart, currBlockSize := readers.CurrentReaders()
+		currentCopies := make([]xio.BlockReader, 0, currLen)
+		for i := 0; i < currLen; i++ {
+			currReader := readers.CurrentReaderAt(i)
+			currCopy, err := s.cloneBlockReaderFn(currReader, currStart, currBlockSize)
+			if err != nil {
+				return nil, err
+			}
+			currentCopies = append(currentCopies, currCopy)
+		}
+		blockReaderCopies = append(blockReaderCopies, currentCopies)
+	}
+
+	// reset the reader to the original state
+	readers.RewindToIndex(initIdx)
+
+	// TODO: pool this type
+	sliceOfSlicesIter := xio.NewReaderSliceOfSlicesFromBlockReadersIterator(blockReaderCopies)
+	multiReaderIterCopy := s.pools.MultiReaderIterator().Get()
+	multiReaderIterCopy.ResetSliceOfSlices(sliceOfSlicesIter, iter.Schema())
+
+	return multiReaderIterCopy, nil
+}
+
+type cloneBlockReaderFn func(xio.BlockReader, xtime.UnixNano, time.Duration) (xio.BlockReader, error)
+
+func (s shallowCloner) cloneBlockReader(
+	reader xio.BlockReader,
+	start xtime.UnixNano,
+	blockSize time.Duration,
+) (xio.BlockReader, error) {
+	// nb: we cannot rely on the reader.Segment.Clone() method as that performs a deep copy.
+	// i.e. it copies the underlying data []byte as well. The copy we provide only
+	// copies the wrapper constructs, it still points at the same []byte as the original.
+
+	seg, err := reader.Segment()
+	if err != nil {
+		return xio.BlockReader{}, err
+	}
+
+	head := s.pools.CheckedBytesWrapper().Get(seg.Head.Bytes())
+	tail := s.pools.CheckedBytesWrapper().Get(seg.Tail.Bytes())
+	checksum := seg.CalculateChecksum()
+	segCopy := ts.NewSegment(head, tail, checksum, ts.FinalizeNone)
```

Comment:
I would throw a comment on here about the FinalizeNone

Reply:
fair

Reply:
done